### ISSUE TITLE
Torch.equal(t1, t2) slower than torch.eq(t1, t2).all() on GPU

### ISSUE THREAD

**Issue by CedricPicron** [pytorch/pytorch#44707]
## 🐛 Bug

`torch.equal(t1, t2)` is slower than its alternative `torch.eq(t1, t2).all()` on GPU. The bug manifests itself after doing "some computation" (as for example self-attention). The only difference between both alternatives is that `torch.equal(t1, t2)` checks the sizes before proceeding and therefore works for all possible `t1` and `t2`, while `torch.eq(t1, t2).all()` only works for non-empty `t1` and `t2` of the same size. 

## To Reproduce
``` python
import copy
import timeit
import torch
from torch import nn

# Before self-attention (some computation)
c = torch.randn(1024, 2, 256, device='cuda')
d = torch.randn(1024, 2, 256, device='cuda')

t1 = timeit.timeit('torch.eq(c, d).all()', number=1, globals=globals())*1e3
t2 = timeit.timeit('torch.eq(c, d).all()', number=1, globals=globals())*1e3
t3 = timeit.timeit('torch.eq(c, d).all()', number=1, globals=globals())*1e3

print(f"First time with eq/all (before self-attention): {t1: .4f} ms")
print(f"Second time with eq/all (before self-attention): {t2: .4f} ms")
print(f"Third time with eq/all (before self-attention): {t3: .4f} ms\n")

t1 = timeit.timeit('torch.equal(c, d)', number=1, globals=globals())*1e3
t2 = timeit.timeit('torch.equal(c, d)', number=1, globals=globals())*1e3
t3 = timeit.timeit('torch.equal(c, d)', number=1, globals=globals())*1e3

print(f"First time with equal (before self-attention): {t1: .4f} ms")
print(f"Second time with equal (before self-attention): {t2: .4f} ms")
print(f"Third time with equal (before self-attention): {t3: .4f} ms\n")

# Perform self-attention (some computation)
a = torch.randn(1024, 2, 256, device='cuda')
b = torch.randn(1024, 2, 256, device='cuda')

self_attention = nn.MultiheadAttention(256, 8).to('cuda')
layers = nn.ModuleList([copy.deepcopy(self_attention) for _ in range(20)])

for layer in layers:
    a = layer(a+b, a+b, a)[0]

# After self-attention (some computation)
t1 = timeit.timeit('torch.eq(c, d).all()', number=1, globals=globals())*1e3
t2 = timeit.timeit('torch.eq(c, d).all()', number=1, globals=globals())*1e3
t3 = timeit.timeit('torch.eq(c, d).all()', number=1, globals=globals())*1e3

print(f"First time with eq/all (after self-attention): {t1: .4f} ms")
print(f"Second time with eq/all (after self-attention): {t2: .4f} ms")
print(f"Third time with eq/all (after self-attention): {t3: .4f} ms\n")

t1 = timeit.timeit('torch.equal(c, d)', number=1, globals=globals())*1e3
t2 = timeit.timeit('torch.equal(c, d)', number=1, globals=globals())*1e3
t3 = timeit.timeit('torch.equal(c, d)', number=1, globals=globals())*1e3

print(f"First time with equal (after self-attention): {t1: .4f} ms")
print(f"Second time with equal (after self-attention): {t2: .4f} ms")
print(f"Third time with equal (after self-attention): {t3: .4f} ms\n")

```

## Results
``` text
First time with eq/all (before self-attention):  0.4065 ms
Second time with eq/all (before self-attention):  0.0777 ms
Third time with eq/all (before self-attention):  0.0763 ms

First time with equal (before self-attention):  0.1580 ms
Second time with equal (before self-attention):  0.1254 ms
Third time with equal (before self-attention):  0.1164 ms

First time with eq/all (after self-attention):  0.0825 ms
Second time with eq/all (after self-attention):  0.0775 ms
Third time with eq/all (after self-attention):  0.0776 ms

First time with equal (after self-attention):  4.1039 ms
Second time with equal (after self-attention):  0.1285 ms
Third time with equal (after self-attention):  0.1273 ms
```

## Interpretation
Before the self-attention computation, all is fine and both alternatives perform relatively well. However, after the self-attention computation, the first `torch.equal` call is much slower than usual. The `torch.equal(t1, t2).all()` does not exhibit such significant slowdown. 

## Expected behavior
It's expected that `torch.equal(t1, t2)` takes similar time as its alternative `torch.eq(t1, t2).all()`.

## Environment
``` text
Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti
Nvidia driver version: 450.66
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.7.0.dev20200903
[pip3] torchvision==0.8.0.dev20200903
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.1.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.1           py38hbc911f0_0  
[conda] numpy-base                1.19.1           py38hfa32c7d_0  
[conda] pytorch                   1.7.0.dev20200903 py3.8_cuda10.2.89_cudnn7.6.5_0    pytorch-nightly
[conda] torchvision               0.8.0.dev20200903      py38_cu102    pytorch-nightly
```

## Additional context
1. See https://discuss.pytorch.org/t/slow-torch-equal-on-gpu-bottleneck-of-nn-multiheadattention/96340 for additional context. Additional results are provided showing that this slow `torch.equal` can form the bottleneck for the `nn.MultiheadAttention` module.

2. Note that the `equal` implementation was recently moved from THC to ATen in PyTorch 1.6 (see https://github.com/pytorch/pytorch/pull/36483). However, when running with PyTorch 1.5, I got similar timings.

3. I also tested this on a different GPU (GTX 1660) with similar findings.

cc @VitalyFedyunin @ngimel

**Comment by CedricPicron** [comment]
### Update
Above statement that `torch.equal(t1, t2)` is slower than its alternative `torch.eq(t1, t2).all()` on GPU after "some computation" might be incomplete or even wrong. 

As explained in https://discuss.pytorch.org/t/slow-torch-equal-on-gpu-bottleneck-of-nn-multiheadattention/96340, `nn.MultiheadAttention` contains some `torch.equal` checks. By profiling the code, I could see that most time was spent in the `torch.equal` call. Subsequently I tried above alternative based on `torch.eq(t1, t2).all()` which seems to be faster according to `timeit.timeit`.

Yet when I adapt https://github.com/pytorch/pytorch/blob/7036e91abd6ba5e9d20afa957913a0cd7c08be81/torch/nn/functional.py#L4095 as follows
``` python
    head_dim = embed_dim // num_heads
    assert head_dim * num_heads == embed_dim, "embed_dim must be divisible by num_heads"
    scaling = float(head_dim) ** -0.5

    # Use eq/all
    bool1 = torch.eq(query, key).all() if query.shape == key.shape else False
    bool2 = torch.eq(key, value).all() if key.shape == value.shape else False

    # Use equal
    # bool1 = torch.equal(query, key)
    # bool2 = torch.equal(key, value)

    # Skip checks
    # bool1 = False
    # bool2 = False

    if not use_separate_proj_weight:
        if bool1 and bool2:
            # self-attention
            q, k, v = linear(query, in_proj_weight, in_proj_bias).chunk(3, dim=-1)

        elif bool2:
            # encoder-decoder attention
            # This is inline in_proj function with in_proj_weight and in_proj_bias
            _b = in_proj_bias
            _start = 0
            _end = embed_dim
            _w = in_proj_weight[_start:_end, :]
```
no speed-up is observed compared to the `torch.equal` variant. The cProfiler gives a different report (no more bottleneck in `equal`), but with same time spent into `multi_head_attention_forward`. This might be consistent with the second warning mentioned in https://pytorch.org/docs/stable/bottleneck.html.

Note that when no check is performed, there is a speed-up. As the user knows which tensors are equal, the user could provide this information to  the forward  method of `nn.MultiheadAttention` instead. This is off-topic and could be discussed in a separate issue.

### Experiment code and results
For completeness, I add the experiment code I used and the obtained results:
``` python
import copy
import timeit
import torch
from torch import nn

a = torch.randn(1024, 2, 256, device='cuda')
b = torch.randn(1024, 2, 256, device='cuda')
_ = torch.matmul(a.transpose(1, 2), b)

self_attention = nn.MultiheadAttention(256, 8).to('cuda')
layers = nn.ModuleList([copy.deepcopy(self_attention) for _ in range(20)])


def run_test(a, b, layers):
    for layer in layers:
        a = layer(a+b, a+b, a)[0]


t = timeit.timeit('run_test(a, b, layers)', number=1, globals=globals())*1e3
print(f"Recorded timeit time: {t: .4f} ms") 
```
By running the code for each of the three variants I got:
```
Eq/all variant: 125.7474 ms
Equal variant: 118.5710 ms
Skip check variant: 34.1028 ms
```

**Comment by ngimel** [comment]
You cannot time cuda operations with timeit, please use pytorch timing utilities https://github.com/pytorch/pytorch/tree/master/torch/utils/_benchmark. You multihead attention benchmarks show that eq/all is not faster than equal, so I don't see what exactly the issue is.

**Comment by ak-source** [comment]
Is there a usecase where that structural equality check is necessary with attention? If not, maybe object id checks like "key is value" could be used instead? Alternatively, an optional argument to manually select attention mode could be added to skip runtime mode detection.

**Comment by CedricPicron** [comment]
Thank you for looking at it @ngimel . I've rerun above benchmarks with the pytorch timing utility you mentioned. For the original experiments, it indeed confirms the updated believe that both equal and the eq/all alternative are equally fast. 

For the subsequent `nn.MultiheadAttention` experiments, it gives all three variants with similar timings
``` text
Eq/all variant: 124.5819 ms
Equal variant: 124.2137 ms
Skip check variant: 116.8404 ms
```
which also settles this matter. 

Thank you for proposing your solutions @ak-source , but I guess it doesn't matter that much anymore. If desired, one could roughly still gain 8 ms over the course of 20 self-attention layers (for a GTX 1050) by skipping these checks. Sorry for dragging you into this.

Finally, before closing this issue, @ngimel, which profiler gives reliable estimates when using CUDA? I used `torch.utils.bottleneck` as described in https://pytorch.org/docs/stable/bottleneck.html?highlight=profiling, but this confirmed (via `torch.autograd.profiler.profile` with  `use_cuda=True`) the wrong belief that `at::equal` was the slowest operation. The mentioned timings at the end are not always correct either (different from the timings obtained with the pytorch timing utility) Is `torch.autograd.profiler.emit_nvtx` a reliable option instead?

**Comment by ngimel** [comment]
Hm, `torch.autograd.profiler.profile` with `use_cuda=True` usually gives accurate measurements, can you show an example where it doesn't? We should fix it in this case. `torch.autograd.profiler.emit_nvtx` won't do anything unless you run under nvprof and then open the produced profile with nvvp (or run with nsight systems/open with nsight systems), which is more onerous than needed for simple timings. 
So, tl;dr, torch.autograd.profile.profile` should work, if you discover a case where it doesn't can you please file an issue?

**Comment by CedricPicron** [comment]
Ok, I will create a new issue with the case I had where  `torch.autograd.profile.profile` was wrong (or inconsistent with the pytorch timing utility). I will do this tomorrow. Thanks for the great help @ngimel !
